rasa/nlu/classifiers/diet_classifier.py

```diff
@@ -1252,6 +1258,193 @@ def __init__(
+        new_sizes = new_feature_sizes[feature_type]
+        if old_sizes != new_sizes:
+            new_layer = self._create_dense_for_sparse_layer(
+                dense_layer=layer_dict["layer"],
+                old_sizes=old_sizes,
+                new_sizes=new_sizes,
+                reg_lambda=self.config[REGULARIZATION_CONSTANT],
+            )
+            layers = self._tf_layers
+            for index, name in enumerate(layer_dict["path"]):
+                if index == len(layer_dict["path"]) - 1:
+                    layers[name] = new_layer
+                else:
+                    layers = layers[name]._tf_layers
+            return True
+        return False
+
+    def _compile_and_fit(self, data_example: Dict[Text, List[FeatureArray]]) -> None:
+        """Compiles modified model and fits a sample data on it.
+
+        Args:
+            data_example: a data example that is stored in DIETClassifier class
+        """
+        self.compile(optimizer=tf.keras.optimizers.Adam(self.config[LEARNING_RATE]))
+        label_key = LABEL_KEY if self.config[INTENT_CLASSIFICATION] else None
+        label_sub_key = LABEL_SUB_KEY if self.config[INTENT_CLASSIFICATION] else None
+
+        model_data = RasaModelData(
+            label_key=label_key, label_sub_key=label_sub_key, data=data_example
+        )
+        self._update_data_signatures(model_data)
+        data_generator = RasaBatchDataGenerator(model_data, batch_size=1)
+        self.fit(data_generator, verbose=False)
+
+    def _find_dense_for_sparse_layer(self, layer: Any, path: List[Text]) -> List[Dict]:
+        """Finds DenseForSparse layers recursively.
+
+        For each layer returns its path in the layers dictionary
+        so that we're able to replace it.
+        """
+        layers = []
+        if hasattr(layer, "_tf_layers"):
+            for name, sub_layer in layer._tf_layers.items():
+                layers += self._find_dense_for_sparse_layer(sub_layer, path + [name])
+        else:
+            if hasattr(layer, "name") and "sparse_to_dense" in layer.name:
+                feature_type = layer.get_feature_type()
+                layers.append(
+                    {"layer": layer, "feature_type": feature_type, "path": path}
+                )
+        return layers
+
+    def _get_dense_for_sparse_layers(self, attribute: Text) -> List[Dict]:
+        """Returns DenseForSparse layers for a given attribute.
+
+        Finds DenseForSparse layers that need to be adjusted for fine-tuning.
+        At this point works for just attribute - text.
+        A layer dictionary could look like this:
+        {'layer': DenseForSparse object, 'feature_type': FEATURE_TYPE_SEQUENCE,
+        'path': ['feature_combining', 'sparse_dense.sequence', 'sparse_to_dense']}.
+
+        Args:
+            attribute: attribute to consider when getting the layers.
+
+        Returns:
+            A List of Dictionary containing layer data.
+        """
+        layers = []
```

Comment:
I think for one attribute there can only be maximum 2 layers that you return - one for sequence, another for sentence. So, a better way to structure `layers` would be - 
```
layers = {
    'SEQUENCE`: {
      'layer': DenseForSparseObject,
      'path_in_model_graph': ...
      },
       'SENTENCE`: {
      'layer': DenseForSparseObject,
      'path_in_model_graph': ...
      },
}
```
Presence of `SEQUENCE` or `SENTENCE` will depend on whether layers for them are actually present or not.